File: KotlinSubmarine/src/main/kotlin/day16/PacketDecoder.kt
package day16

import Puzzle
import java.io.File

fun main() {
    PacketDecoder().run()
}

// https://adventofcode.com/2021/day/16

class PacketDecoder : Puzzle(16, "Packet Decoder") {
    override val puzzleInput = "day16_input.txt"
    override val testInput = arrayOf(
        "day16_test1.txt",
        "day16_test2.txt",
        "day16_test3.txt",
        "day16_test4.txt",
        "day16_test5.txt",
        "day16_test6.txt",
        "day16_test7.txt")
    override val label1 = "version sum"
    override val label2 = "packet value"

    private lateinit var packet: Packet

    override fun part1(input: String): Long {
        packet = decode(File(input).readText())
        return versionSum
    }
    override fun part2(input: String): Long {
        return packet.value
    }
}

private var versionSum = 0L

private fun decode(hex: String): Packet {
    //println(hex)
    versionSum = 0L
    val reader = BitReader(hex)
    return decodePacket(reader, 0)
}

private fun decodePacket(reader: BitReader,
                         depth: Int): Packet {

    // Depth and indent is only used for debugging print statements
    var indent = ""
    for (i in 1..depth) indent += "  "

    val packetVersion = reader.readBits(3)
    val packetType = reader.readBits(3).toInt()
    val packet = Packet()
    packet.size = 6
    packet.value = 0

    versionSum += packetVersion

    if (packetType == 4) {
        // literal packets
        var number = 0L
        var more = 1L
        while (more != 0L) {
            more = reader.readBits(1)
            val n = reader.readBits(4)
            number = (number shl 4) + n
            packet.value = number
            packet.size += 5
        }
        //println("${indent}v $packetVersion literal $number")
    } else {
        // operator packets
        val lenType = reader.readBits(1).toInt()
        packet.size++

        //val names = arrayOf("sum", "product", "minimum", "maximum", "literal", "greater-than", "less-than", "equals")
        //val op = names[packetType]
        when (lenType) {
            0 -> {
                var packetLength = reader.readBits(15)
                packet.size += 15
                //println("${indent}v $packetVersion $op sub-packet bits $packetLength")
                while (packetLength > 0) {
                    val subPacket = decodePacket(reader, depth+1)
                    packet.subPackets.add(subPacket)
                    packet.size += subPacket.size
                    packetLength -= subPacket.size
                }
            }
            1 -> {
                val count = reader.readBits(11)
                packet.size += 11
                //println("${indent}v$packetVersion $op sub-packet count $count")
                for (i in 1 .. count) {
                    val subPacket = decodePacket(reader, depth+1)
                    packet.subPackets.add(subPacket)
                    packet.size += subPacket.size
                }
            }
        }

        // Probably could condense this with functions
        when (packetType) {
            0 -> { // sum
                var sum = 0L
                packet.subPackets.forEach { sum += it.value }
                packet.value = sum
            }
            1 -> { // product
                if (packet.subPackets.size > 0) {
                    var product = 1L
                    packet.subPackets.forEach { product *= it.value }
                    packet.value = product
                }
            }
            2 -> { // minimum
                var min = Long.MAX_VALUE
                packet.subPackets.forEach {
                    if (it.value < min)
                        min = it.value
                }
                packet.value = min
            }
            3 -> { // maximum
                var max = Long.MIN_VALUE
                packet.subPackets.forEach {
                    if (it.value > max)
                        max = it.value
                }
                packet.value = max
            }
            5 -> { // greater-than
                if (packet.subPackets[0].value > packet.subPackets[1].value)
                    packet.value = 1
                else
                    packet.value = 0
            }
            6 -> { // less-than
                if (packet.subPackets[0].value < packet.subPackets[1].value)
                    packet.value = 1
                else
                    packet.value = 0
            }
            7 -> { // equals
                if (packet.subPackets[0].value == packet.subPackets[1].value)
                    packet.value = 1
                else
                    packet.value = 0
            }
        }
        //println("${indent}value: ${packet.value}")
    }
    return packet
}

// Didn't really want to make a Packet class but here we are
private class Packet() {
    val subPackets = mutableListOf<Packet>()
    var size: Int = 0
    var value: Long = 0
}

// Reads arbitrary bit length numbers from a hexidecimal string input
private class BitReader(private val hexStream: String) {
    private var hexIndex = 0
    private var bitIndex = 0
    // Seems like there should be a language feature for this:
    private val hexMap = mapOf(
        '0' to 0,
        '1' to 1,
        '2' to 2,
        '3' to 3,
        '4' to 4,
        '5' to 5,
        '6' to 6,
        '7' to 7,
        '8' to 8,
        '9' to 9,
        'A' to 10,
        'B' to 11,
        'C' to 12,
        'D' to 13,
        'E' to 14,
        'F' to 15)
    private val bitMap = mapOf(
        0 to 0b1000,
        1 to 0b0100,
        2 to 0b0010,
        3 to 0b0001)

    fun readBits(bits: Int): Long {
        var output = 0L
        var input = current()
        for (i in 0 until bits) {
            output = output shl 1
            val mask = bitMap[bitIndex]!!
            if ((input and mask) != 0)
                output += 1
            bitIndex++
            if (bitIndex > 3) {
                bitIndex = 0
                hexIndex++
                if (hexIndex >= hexStream.length) {
                    return output
                }
                input = current()
            }
        }
        return output
    }

    private fun current(): Int {
        return hexMap[hexStream[hexIndex]]!!
    }
}

File: GoSubmarine/day16/packetDecoder.go
package main

import (
	"math"

	"github.com/tkrehbiel/adventofcode2021/gosubmarine/common"
)

// https://adventofcode.com/2021/day/16
// Packet Decoder

/*
This is a Golang version of my Kotlin solution.
I braced myself for another monster, but this one was a fairly straightforward
decoding of a binary stream, like the good old days. I made a serial binary reader
class to simplify the parsing, the rest was typing out the complicated rules in code
and debugging the problems as they occurred.
*/

func main() {
	common.Run(common.PuzzleConfig{
		Puzzle:      new(packetDecoder),
		Day:         16,
		Title:       "Packet Decoder",
		PuzzleInput: "day16_input.txt",
		TestInput: []string{
			"day16_test1.txt",
			"day16_test2.txt",
			"day16_test3.txt",
			"day16_test4.txt",
			"day16_test5.txt",
			"day16_test6.txt",
			"day16_test7.txt",
		},
	})
}

type packetDecoder struct {
	versionSum int
	packet     packet
}

type packet struct {
	value    int
	size     int
	children []packet
}

func (p *packetDecoder) Common(input string) {}

// Part1 returns the first puzzle result
func (p *packetDecoder) Part1(input string) (result int) {
	p.versionSum = 0
	reader := newBitReader(common.ReadText(input))
	p.packet = getPacket(p, &reader)
	return p.versionSum
}

// Part2 returns the second puzzle result
func (p *packetDecoder) Part2(input string) (result int) {
	// Already calculated in Part1 above:
	return p.packet.value
}

// getPacket decodes the next packet and all of its sub-packets from the bit reader
func getPacket(puzzle *packetDecoder, reader *bitReader) (p packet) {
	version := reader.read(3)
	packetType := reader.read(3)
	p.size = 6

	puzzle.versionSum += version

	if packetType == 4 {
		// literal packet
		more := true
		for more {
			more = reader.read(1) != 0
			decimal := reader.read(4)
			p.value = p.value<<4 | decimal
			p.size += 5
		}
	} else {
		// operator packet
		lenType := reader.read(1)
		p.size++
		if lenType == 0 {
			length := reader.read(15)
			p.size += 15
			for length > 0 {
				sub := getPacket(puzzle, reader)
				p.size += sub.size
				p.children = append(p.children, sub)
				length -= sub.size
			}
		} else {
			count := reader.read(11)
			p.size += 11
			for count > 0 {
				sub := getPacket(puzzle, reader)
				p.size += sub.size
				p.children = append(p.children, sub)
				count--
			}
		}
		operator(packetType, &p)
	}
	return p
}

// operator computes the operator value from the subpackets
func operator(packetType int, op *packet) {
	op.value = 0
	switch packetType {
	case 0:
		for _, p := range op.children {
			op.value += p.value
		}
	case 1:
		product := 1
		for _, p := range op.children {
			product *= p.value
		}
		op.value = product
	case 2:
		min := math.MaxInt
		for _, p := range op.children {
			if p.value < min {
				min = p.value
			}
		}
		op.value = min
	case 3:
		max := math.MinInt
		for _, p := range op.children {
			if p.value > max {
				max = p.value
			}
		}
		op.value = max
	case 5:
		if op.children[0].value > op.children[1].value {
			op.value = 1
		}
	case 6:
		if op.children[0].value < op.children[1].value {
			op.value = 1
		}
	case 7:
		if op.children[0].value == op.children[1].value {
			op.value = 1
		}
	}
}
